cupyx/scipy/sparse/_index.py

```diff
@@ -0,0 +1,388 @@
+"""Indexing mixin for sparse matrix classes.
+"""
+
+import cupy
+from cupy import core
+
+import numpy
+
+
+_int_scalar_types = (int, numpy.integer)
+
+
+def _get_csr_submatrix(Ap, Aj, Ax,
+                       start_maj, stop_maj,
+                       start_min, stop_min):
+    """Return a submatrix of the input sparse matrix by
+    slicing both major and minor axes.
+
+    Args
+        Ap : indptr array from input sparse matrix
+        Aj : indices array from input sparse matrix
+        Ax : data array from input sparse matrix
+        start_maj : starting index of major axis
+        stop_maj : ending index of major axis
+        start_min : starting index of minor axis
+        stop_min : ending index of minor axis
+
+    Returns
+        Bp : indptr array of output sparse matrix
+        Bj : indices array of output sparse matrix
+        Bx : data array of output sparse matrix
+    """
+
+    Bp = Ap[start_maj:stop_maj+1]
+
+    start_offset = Bp[0]
+    stop_offset = Bp[-1]
+
+    Aj = Aj[start_offset:stop_offset]
+    Ax = Ax[start_offset:stop_offset]
+
+    Aj_copy = cupy.empty(Aj.size+1, dtype=Aj.dtype)
+    Aj_copy[1:] = Aj
+
+    Aj_copy[Aj_copy < start_min] = -1
+    Aj_copy[Aj_copy >= stop_min] = -1
+
+    mask = Aj_copy > -1
+
+    Aj_copy[mask] = 1
+    Aj_copy[~mask] = 0
+
+    cupy.cumsum(Aj_copy, out=Aj_copy)
+
+    Aj_copy -= Aj_copy[0]
+
+    Bp = Aj_copy[Bp-start_offset]
+    Bj = Aj[mask[1:]] - start_min
+    Bx = Ax[mask[1:]]
+
+    return Bp, Bj, Bx
+
+
+_set_boolean_mask_for_offsets = core.ElementwiseKernel(
+    'raw T start_offsets, raw T stop_offsets', 'raw bool mask',
+    '''
+    for (int jj = start_offsets[i]; jj < stop_offsets[i]; jj++) {
+        mask[jj] = true;
+    }
+    ''', 'set_boolean_mask_for_offsets', no_return=True)
+
+
+def _csr_row_slice(start, step, Ap, Aj, Ax, Bp):
```

Comment:
How about renaming to `start_maj` and `step_maj`?
```suggestion
def _csr_row_slice(start_maj, step_maj, Ap, Aj, Ax, Bp):
```